### ISSUE TITLE
aiomysql closed exception

### ISSUE THREAD

**Issue by watsy0007** [tortoise/tortoise-orm#127]
I don't know how to catch this error. and reconnect
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sanic/app.py", line 917, in handle_request
    response = await response
  File "/app/blueprints/exchanges/exchange.py", line 23, in exchanges
    request['page'], request['per_page'])
  File "/app/services/exchange.py", line 13, in query_exchanges
    count = await query.count()
  File "/usr/local/lib/python3.7/site-packages/tortoise/queryset.py", line 531, in _execute
    result = await self._db.execute_query(str(self.query))
  File "/usr/local/lib/python3.7/site-packages/tortoise/backends/mysql/client.py", line 23, in wrapped
    return await func(self, query, *args)
  File "/usr/local/lib/python3.7/site-packages/tortoise/backends/mysql/client.py", line 124, in execute_query
    await cursor.execute(query)
  File "/usr/local/lib/python3.7/site-packages/aiomysql/cursors.py", line 239, in execute
    await self._query(query)
  File "/usr/local/lib/python3.7/site-packages/aiomysql/cursors.py", line 457, in _query
    await conn.query(q)
  File "/usr/local/lib/python3.7/site-packages/aiomysql/connection.py", line 427, in query
    await self._execute_command(COMMAND.COM_QUERY, sql)
  File "/usr/local/lib/python3.7/site-packages/aiomysql/connection.py", line 663, in _execute_command
    self._write_bytes(prelude + sql[:chunk_size - 1])
  File "/usr/local/lib/python3.7/site-packages/aiomysql/connection.py", line 608, in _write_bytes
    return self._writer.write(data)
  File "/usr/local/lib/python3.7/asyncio/streams.py", line 305, in write
    self._transport.write(data)
  File "uvloop/handles/stream.pyx", line 671, in uvloop.loop.UVStream.write
  File "uvloop/handles/handle.pyx", line 159, in uvloop.loop.UVHandle._ensure_alive
RuntimeError: unable to perform operation on <TCPTransport closed=True reading=False 0x55931a65f778>; the handler is closed
```

**Comment by grigi** [comment]
Hi @watsy0007 Thank you for reporting this issue.

This looks to be related to #117 where, for example, MySQL server auto disconnects idle connections, and Tortoise doesn't have an auto-reconnect functionality at this stage.

aiopg has _some_ mitigation built it, but not entierly adequate. So we should handle this as more important.

**Comment by watsy0007** [comment]
@grigi > . <
I use tortoise-orm in production now.
if can't reconnect, services are unavailable...
I want to know when to support this feature? or have a plan?

**Comment by grigi** [comment]
Well, then lets get working!

This is mysql you are using?
Does the exception happen instantly? or does it freeze for a few seconds before raising?

I'm thinking of a simple decorator around the pieces of code that integrates with `aiomysql`. if a connection/tcp error happens, it re-connects, and re-runs the decorated function once.

So just a basic auto-reconnect. It should help.

The question is, what exception to catch, as `RuntimeError` is very broad.
And how do I test this? :thinking:

**Comment by watsy0007** [comment]
yeah.
I use aiomysql.

how about shut down mysql service?

**Comment by grigi** [comment]
I have a simple reconnection wrapper, and busy testing it now with the quart example.
Right now, It does successfully reconnect but the error leaks back up, so the server does a 500 for one request, then works again afterwards.

I think it is because we do an `asyncio.gather()` to force multiple ORM events concurrently.
So somewhere the error is bubbling up. without that it doesn't fail.
I'm still tracing, and once the manual solution works satisfactorily I'll ask you to test it out and see if it resolves your problem?

**Comment by watsy0007** [comment]
I'm now sure.
in my case, I use redis as cache with tortoise-orm.
now i must clear the cache everywhere, beacuse I don't known disconn when did disconnect happend. 

I can use it in the short term.

**Comment by grigi** [comment]
I got my test-case working. I was not awaiting the lock :man_facepalming: 

I released v0.11.9, please test if it resolves your issue?

**Comment by watsy0007** [comment]
@grigi thanks.
